### PR TITLE
Dev plugin rabbitmq

### PR DESCRIPTION
**Pull request by sadat-ali-khan**

### CODE REVIEW

**Comment by jiaqifeng** [comment]
@Xylus I've update the rabbitmq plugin code, but checks failed for codecov/patch and codecov/project. What should I do? Does it mean the test code coverage is not enough?

**Comment by Xylus** [comment]
@jiaqifeng The code coverage for plugins don't cover integration tests so the coverage is usually low. That said, if you could add integration tests for rabbitmq plugin in the future, that'd be great (via another PR as this one will merged soon).

File: plugins/rabbitmq/src/main/java/com/navercorp/pinpoint/plugin/rabbitmq/interceptor/RabbitMQPublishInterceptor.java [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
 import com.rabbitmq.client.AMQP;
 
+import java.util.HashMap;
+import java.util.Map;
+
 /**
  * @author Jinkai.Ma
+ * @author Jiaqi Feng
  */
 @Scope(value = RabbitMQConstants.RABBITMQ_SCOPE)
```
Comment (Xylus) [inline plugins/rabbitmq/src/main/java/com/navercorp/pinpoint/plugin/rabbitmq/interceptor/RabbitMQPublishInterceptor.java]: Annotations have been prohibited in interceptors due to issues present in #2862 
Could you remove `@Scope` and use `InstrumentMethod.addScopedInterceptor(...)` method instead?

**Comment by jiaqifeng** [comment]
@Xylus I have added the integration test and remove the scope annotation. I just committed on same branch, you could cherry-pick it. Or I can commit it on a new branch and reset this branch if needed.
Using qpid as amqp server really took my many time for so less info on internet.

File: plugins/rabbitmq/src/main/java/com/navercorp/pinpoint/plugin/rabbitmq/interceptor/RabbitMQPublishInterceptor.java (hunk 1 above)
Comment (jiaqifeng) [reply plugins/rabbitmq/src/main/java/com/navercorp/pinpoint/plugin/rabbitmq/interceptor/RabbitMQPublishInterceptor.java]: Done.

**Comment by jiaqifeng** [comment]
@Xylus  #3495 seems just need to intercepte a new class. Have you fix it?

File: plugins/rabbitmq/src/main/java/com/navercorp/pinpoint/plugin/rabbitmq/interceptor/RabbitMQPublishInterceptor.java (hunk 1 above)
Comment (Xylus) [reply plugins/rabbitmq/src/main/java/com/navercorp/pinpoint/plugin/rabbitmq/interceptor/RabbitMQPublishInterceptor.java]: :+1:

**Comment by Xylus** [comment]
@jiaqifeng We'll merge it after 1.7.0 releases as the code is currently frozen in preparation for the release.
Thanks!

**Comment by Xylus** [comment]
@jiaqifeng pushing to your feature branch to resolve conflict

**Comment by jiaqifeng** [comment]
@Xylus 
It seems release 1.7's changes split with the HEAD when I rebased this branch. And I will rebase it later since the next release will be after monthes.

**Comment by Xylus** [comment]
@jiaqifeng Yeh, we had the history rewritten unfortunately :(
The simplest way would probably be to delete your local repository's *master* branch and get a new copy of it from naver/pinpoint's *master*. Then create a new feature branch for rabbitmq plugin, and just cherry-pick the commits from your *dev_plugin_rabbitmq* branch.
Though you might want to wait until after 1.7.1 releases this week (a simple bug patch).
Thanks again.

**Comment by jiaqifeng** [comment]
@Xylus 
I will create a temp branch, cherry-pick the changes and reset --hard dev_plugin_rabbitmq. Using git is easy to me. I will do it 2 weeks later.

**Comment by Xylus** [comment]
@jiaqifeng Sounds good :)

**Comment by jiaqifeng** [comment]
@Xylus I've rebased the branch on the 1.7.2-SNAPSHOT release and passed all rabbitmq integration tests. Any problem please me know.

**Comment by Xylus** [comment]
@jiaqifeng lgtm, thank you for your contribution!

**Comment by jiaqifeng** [comment]
@Xylus Wow, you are so quick. While today somebody said it could not continue transaction when using springboot amqp. The root cause is spring boot start a new thread to call app's message handler when receive message in receiving thread. I fix it but I think it should be tested more.

**Comment by Xylus** [comment]
@jiaqifeng Yeh, I was going to look into this after merging!
I think it's the same issue as #3495 (ActiveMQ but I think the issue is exactly the same maybe with a different handler).

If you'd like to work on the fix, that would be super!
(Feel free to create a new issue for this if you decide to do it)

**Comment by jiaqifeng** [comment]
@Xylus I will watch on this bug, I think I have fixed it and tested it on specific version of springboot but not tested it on other versions. When I think the fix is stable enough I will PR it.

**Comment by Xylus** [comment]
@jiaqifeng Sounds great!

**Comment by geekMessi** [comment]
@Xylus @jiaqifeng 
Had the bug like #3495 been fixed ?
RabbitMQPlugin RabbitMQConsumeInterceptor losed the trace context at the BlockingQueueConsumer$InternalConsumer#handleDelivery stage.

**Comment by Xylus** [comment]
@geekMessi working on it!

**Comment by Xylus** [comment]
@geekMessi @jiaqifeng Fix has been merged. Please test it out if you can as I've only tried it in a test environment.
It would be great to have your feedback on it.
Thanks.

**Pull request merged**